gax/src/main/java/com/google/api/gax/rpc/ServerStreamingCallable.java

```diff
@@ -212,6 +212,16 @@ public void serverStreamingCall(RequestT request, ApiStreamObserver<ResponseT> r
   public ServerStreamingCallable<RequestT, ResponseT> withDefaultCallContext(
       final ApiCallContext defaultCallContext) {
     return new ServerStreamingCallable<RequestT, ResponseT>() {
+      @Override
```

Comment:
So it is an anonymous internal class in `ServerStreamingCallable` which extends itself (i.e. again `ServerStreamingCallable`) and then uses the original `ServerStreamingCallable.this` directly in its implementation. So it creates its own child, which still depends on itself. Its like [asexual reproduction](https://en.wikipedia.org/wiki/Asexual_reproduction) in Java, but not really, because the child still physically connected to the parentюю. 

Wow...  (gross...). I just hope that you know what you are doing here. Can you please at least provide a comment why this is done. Is it to fix the problems caused by callables chaining? Even if this solves the actual problem, won't it basically make a "hole"  in the chain of callables (which later may cause unpredictable results and lots of frustration)?

Reply:
I agree this is a bad pattern, but it's something that already exists:

https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/UnaryCallable.java#L121-L128

https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/ServerStreamingCallable.java#L214-L222

https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/BidiStreamingCallable.java#L250-L259

I'm not sure what you mean about having a hole in chain?
Say you have a server streaming callable called SS1, calling SS1.first() will return a unary callable U1. If you invoke U1.call(), it will end up calling SS1.call() and then cancel the stream.

Without the changes in this pr, if invoked SS1.withDefaultCallContext(), you would get new anonymous server streaming callable, SS2. When you invoked SS2.first(), it will return a new unary callable called U2. Which when invoked, would call SS2.call(), which in turn calls SS1.call(), bypassing U1.

The only thing this PR does, it that it forces the call through U1. This way we only bypass SS2.call(), which is fine. The net result is that if the next callable has a customized implementation of first() we respect it.

I added a comment to try to clarify in the code

Reply:
Ok, I did not realize that we already had this pattern in our codebase. LGTM.